File: neet_code_all_cpp/sliding_window/freq_of_most_frequent_element.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>

// 模範解答
class SolutionAns {
    public:
        int maxFrequency(std::vector<int> nums, int k) {
            std::sort(nums.begin(), nums.end());

            auto [left, right, res] = std::tuple(0, 0, 0);
            long long total = 0;
            int n = nums.size();

            while(right < n) {
                total += nums[right];

                while((long)(right - left + 1) * nums[right] > total + k) {
                    total -= nums[left];
                    ++left;
                }

                res = std::max(res, right - left + 1);
                ++right;
            }

            return res;
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, int>;

    Case case_1 = {{1, 2, 4}, 5};
    // => 3
    Case case_2 = {{1, 4, 8, 13}, 5};
    // => 2
    Case case_3 = {{3, 9, 6}, 2};
    // => 1

    SolutionAns s_ans;

    std::cout << s_ans.maxFrequency(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.maxFrequency(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.maxFrequency(case_3.first, case_3.second) << std::endl;
}

File: neet_code_all/src/sliding_window/freq_of_most_frequent_element.rs
// 解けなかった(全く方針が立たなかった)
struct Solution {}
impl Solution {
    pub fn max_frequency(nums: Vec<i32>, k: i32) -> i32 {
        let mut nums = nums;
        let n = nums.len();
        //nums.sort();

        let mut result = 0;
        for i in 0..n {}

        result
    }
}

// 模範解答 => WAとなって正しくない
struct SolutionAns {}
impl SolutionAns {
    pub fn max_frequency(nums: Vec<i32>, k: i32) -> i32 {
        let mut nums = nums;
        let (mut left, mut right, mut res, mut total) = (0, 0, 0, 0);

        nums.sort_unstable();

        while right < nums.len() {
            total += nums[right];

            while (nums[right] * (right - left + 1) as i32) - total > k {
                total -= nums[left];
                left += 1;
            }

            res = res.max(right - left + 1);
            right += 1;
        }

        res as i32
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn max_frequency(nums: Vec<i32>, k: i32) -> i32 {
        let k = k as i64;
        let mut nums = nums;
        nums.sort();

        let (mut left, mut right) = (0, 0);
        let mut total = 0_i64;
        let mut result = 0;

        while right < nums.len() {
            total += nums[right] as i64;

            while (right - left + 1) as i64 * nums[right] as i64 > total + k {
                total -= nums[left] as i64;
                left += 1;
            }

            result = result.max(right - left + 1);
            right += 1;
        }

        result as i32
    }
}

fn main() {
    let case_1 = (vec![1, 2, 4], 5);
    // => 3
    let case_2 = (vec![1, 4, 8, 13], 5);
    // => 2
    let case_3 = (vec![3, 9, 6], 2);
    // => 1

    /*
    println!(
        "case_1: {}",
        SolutionAns::max_frequency(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAns::max_frequency(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAns::max_frequency(case_3.0.clone(), case_3.1)
    );
    */

    println!(
        "case_1: {}",
        SolutionAnsCpp::max_frequency(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::max_frequency(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::max_frequency(case_3.0.clone(), case_3.1)
    );
}
